apps/src/dance/LottieDancerRenderer.ts

```diff
@@ -0,0 +1,673 @@
+import lottie, {
+  type AnimationItem,
+  type CanvasRendererConfig,
+} from 'lottie-web';
+
+import appConfig from '../music/appConfig';
+
+import {
+  Canvas2D,
+  CanvasAnimConfig,
+  HeadImageInfo,
+  LottieAssetImage,
+  LottieAssetPrecomp,
+  LottieColorNode,
+  LottieImageLayer,
+  LottieJSON,
+  LottieLayer,
+  LottieLayerCommon,
+  LottiePrecompLayer,
+  LottieShapeAny,
+  LottieShapeFillOrStroke,
+  LottieShapeGroup,
+  Palette,
+  RGBA,
+} from './LottieDancerTypes';
+
+// Default assets
+const BASE_HOST = 'https://curriculum.code.org/media/musiclab/generate';
+const ASSETS_FOLDER = 'basic2';
+const TEST_BASE_DANCER = 'duck';
+const TEST_GENERATED_DANCER = 'basic-frog-baseball-cap-00';
+
+// Accessory-name mapping used to recolor vector content in the Lottie JSON.
+const ACCESSORY_MAP = {
+  secondary: new Set<string>([
+    'bracelet',
+    'shirt high',
+    'shirt low',
+    'shirt low 2',
+    'line',
+  ]),
+  tertiary: new Set<string>([
+    'cuff',
+    'hip',
+    'pelvis',
+    'torso accent',
+    'belly',
+    'leg cuff',
+    'arc',
+  ]),
+};
+
+export default class LottieDancerRenderer {
+  // Injected by ExternalDancerLayer
+  private ctx: Canvas2D | null = null;
+
+  private anim: AnimationItem | null = null;
+  private animationData: LottieJSON | null = null;
+
+  private palette: Palette | null = null;
+  private totalFrames: number | null = null;
+
+  // Values pulled from appConfig/localStorage
+  private readonly headScale: number;
+  private readonly assetsPath: string;
+  private readonly skeletonName: string;
+  private readonly variant: string;
+  private readonly dancerName: string;
+  private cachedAnimationData: {[key: string]: LottieJSON} = {};
+
+  // Optionally reported by getCompSize()
+  private compW?: number;
+  private compH?: number;
+
+  constructor() {
+    const localStorageDancer = safeParseJSON(
+      localStorage.getItem('dancer-ai-generate')
+    ) as {
+      adlibOption?: string;
+      choices?: string[];
+      variant?: number | string;
+    } | null;
+
+    const {adlibOption, choices, variant} = localStorageDancer || {};
+
+    const resolvedDancerName =
+      choices && String(variant)
+        ? `${choices.join('-')}-0${String(variant)}`
+        : TEST_GENERATED_DANCER;
+
+    this.headScale = 0.5;
+    this.assetsPath =
+      (appConfig.getValue('path') as string) || adlibOption || ASSETS_FOLDER;
+    this.skeletonName =
+      (appConfig.getValue('skeleton') as string)?.toLowerCase() ||
+      TEST_BASE_DANCER;
+    this.variant = (variant as string) || '00';
+    this.dancerName =
+      (appConfig.getValue('dancer') as string)?.toLowerCase() ||
+      resolvedDancerName;
+    this.cachedAnimationData = {};
+  }
+
+  /** DanceParty hands us a p5.Graphics 2D context. */
```

Comment:
What about other clients, such as Music Lab and the "generate dancer" mode of Dance Lab2?

Reply:
Good point. Will update the comment.